ETK: Fix for dotcom legacy FSE to work with current Gutenberg versions

creativecoder opened the pull request:
#### Changes proposed in this Pull Request

* Update Dotcom legacy FSE to be compatible with the current version of Gutenberg

#### Testing instructions

For ease of testing, I've created D63129-code with these changes along with updating which version of Gutenberg is loaded, for testing on Simple Sites.

**Test block migrations when upgrading Gutenberg**

- Create a Simple site that has legacy FSE enabled
  - Normally this can be done by following the onboarding flow from https://horizon.wordpress.com/start/test-fse/fse-themes. A recent regression prevents us from doing so.
  - The alternative is creating a simple wordpress site, updating a legacy FSE compatible theme, and updating the blog sticker for said simple site.
    - Create a simple site https://wordpress.com/start/domains
    - Open an ssh connection to your dotcom sandbox in a terminal
    - Run this command in the terminal `wp blog-stickers add --who=YOURUSERNAME --url=YOURTESTSITEDOMAIN --sticker=full-site-editing`
    - Run this command in the terminal `wp theme activate "pub/maywood" --url=YOURTESTSITEDOMAIN`
    - Navigate to any page to verify that legacy FSE is enabled
- BEFORE you sandbox the site, import the file in D63129-code that has a post with all the [Core blocks available in Gutenberg v9.7.4](https://github.com/WordPress/gutenberg/tree/release/9.7/packages/block-library/src) 
- Check that the post imported successfully to your site
- Apply D63129-code, sandbox the site, and reload the post in the editor
- Notice any errors in the JS console that may be related to block migrations _for Core blocks_
- Save the post and make sure all blocks display as expected on the front-end

**Test Compatibility of Legacy FSE Features**

- Create a Simple site that has legacy FSE enabled, if you don't have one already (see above for instructions)
- Apply D63129-code in your sandbox, and sandbox both the site domain _AND_ public-api.wordpress.com
- Try editing a page, including changing the header and footer
- Check that the changes save correctly and can be viewed on the front-end of the site

**Test on Atomic Sites**

- The easiest way to test on Atomic sites is the create a Simple Site using the above instructions, purchase a Business plan, and then install a plugin to transfer the site to Atomic. Note that [Legacy FSE features are currently broken on initial transfer](https://github.com/Automattic/wp-calypso/issues/41825)--you'll need to switch the theme, then switch back to Maywood or another legacy FSE compatible theme to scaffold the template content on theme activation.
- Next download a build of the plugin from this branch, and upload to your site for testing, as described in PCYsg-ly5-p2
- Check that you can edit template parts, save the changes, and they show up correctly on the site.

### Deploy note

To deploy this to wpcom, we'll need to combine the ETK changes from this PR with the changes to Gutenberg loading in D63129-code.

Atomic sites will just need the usual ETK release.

Commit:
Initial fix for legacy FSE

Commit:
Fix editor-template-classes plugin by updating block list parent class

Commit:
Correctly get the template part to edit with current Gutenberg version

Commit:
Remove legacy fse filter override (used for testing)

Commit:
Fix a couple style issues with the editor

creativecoder commented on apps/editing-toolkit/editing-toolkit-plugin/dotcom-fse/class-full-site-editing.php:
```diff
@@ -61,6 +61,7 @@ private function __construct() {
 		add_action( 'pre_delete_term', array( $this, 'restrict_template_taxonomy_deletion' ), 10, 2 );
 		add_action( 'transition_post_status', array( $this, 'restrict_template_drafting' ), 10, 3 );
 		add_action( 'admin_menu', array( $this, 'remove_wp_admin_menu_items' ) );
+		add_filter( 'block_editor_rest_api_preload_paths', array( $this, 'preload_template_parts' ), 10, 2 );
```
This filter isn't available until WP 5.8, but preloading isn't required for editing template parts to work.

creativecoder commented on apps/editing-toolkit/editing-toolkit-plugin/dotcom-fse/plugins/editor-template-classes/index.js:
```diff
@@ -30,7 +30,7 @@ const EditorTemplateClasses = withSelect( ( select ) => {
 	return { templateClasses };
 } )( ( { templateClasses } ) => {
 	const blockListInception = setInterval( () => {
-		const blockListParent = document.querySelector( '.block-editor__typewriter > div' );
```
The TypeWriter component was [switched to use hooks](https://github.com/WordPress/gutenberg/pull/26986) which eliminated this class.

jeyip commented:
## Testing
### Requirements
**Test block migrations when upgrading Gutenberg**
- [x] Did not see JS console errors that might have been related to block migrations for Core blocks
- [x] Block migrations display as expected on the front-end. Here's the behavior I verified:
  - Archive: Links to blog archives
  - Audio: Can play and pause music, can change volume, can skip through the provided track
  - Button: Test button click opens new web page
  - Buttons: Buttons open new web page 
  - Calendar: Displays current month and links to blog posts made on specific dates
  - Categories: Links to existing categories
  - Code: Displays formatted code content
  - Columns: Displays multiple columns of content
  - Cover: Displays a cover image overlayed by text
  - Embed: Can play the provided "Deep Focus" video
  - File: Can open pdf file and download it
  - Gallery: Displays collection of images and can enter isolated carousel view of each image
  - Group: Groups the list content + cat image
  - Comments: Displays "No comments to show."
  - Latest posts: Displays latest posts for blog. Links to each one.
  - Media & Text: Displays image horizontally aligned with descriptive text content
  - Navigation: Links to social media sites
  - Preformatted: Displays text content that respects spacing and tabs
  - Pullquote: Displays formatted quote
  - Quote: Displays a different, uniquely formatted quote
  - RSS: Displays entries from RSS
  - Search: Can search for content
  - Separator: Displays dividing line
  - Shortcode: Displays gravatar shortcode
  - Social links: Links to social media
  - Spacer: Displays empty space
  - Table: Displays styled, tabular data
  - Tag cloud: Displays cloud of most used tags
  - Verse: Displays specially formatted poetry
  - Video: Can play video with macho bunny
  - Page break: Separates content into multiple pages

**Simple Sites: Test Compatibility of Legacy FSE Features**
- [x] Header changes can be persisted (modified site description, added text and image content)
- [x] Header changes are displayed on the front end and legacy site editor
- [x] Footer changes can be persisted (modified site description, added text and image content)
- [x] Footer changes are displayed on the front end and legacy site editor

**Atomic Sites: Test Compatibility of Legacy FSE Features**
- [x] Header changes can be persisted (modified site description, added text and image content)
- [x] Header changes are displayed on the front end and legacy site editor
- [x] Footer changes can be persisted (modified site description, added text and image content)
- [x] Footer changes are displayed on the front end and legacy site editor

### Browsers
- [x] Edge
- [x] Firefox
- [x] Safari
- [x] Chrome

Things appear to test well 🎉  To be on the safe side though, it'd be awesome if someone that was around for legacy dotcom FSE could double check functionality as well @Addison-Stavlo, @Copons, @mattwiebe

creativecoder commented:
I added some additional testing instructions for Atomic sites to the issue description. They are already running a recent version of Gutenberg, so the main thing to test is make sure this change fixes legacy FSE template editing.

jeyip commented on apps/editing-toolkit/editing-toolkit-plugin/dotcom-fse/templates/class-rest-templates-controller.php:
```diff
@@ -11,6 +11,28 @@
+	 * @param WP_REST_Request $request API request for template parts.
+	 * @return WP_REST_Response
+	 */
+	public function get_items( $request ) {
```
`NB:`

It's curious to me that we need to override the get_**items**, instead of the get_**item** action, to return a _singular_ `template_part` resource.

I see in [the @see link](https://github.com/WordPress/gutenberg/blob/6c94fb7a233e849e0fff51deffba6ab2866c60f4/packages/edit-post/src/editor.js#L71) in the inline comments that the post editor is bootstrapped via `getEntityRecords` instead of `getEntityRecord`. If I understand correctly, that's the reason why?

The changes were added in [this PR](https://github.com/WordPress/gutenberg/pull/27910/files#diff-d00cfe048041886ee9b16f6f42422317b6e770cbea20cbe0e4d0ca7c8e0ec0e1), which was a huge undertaking, so I wonder if it was an oversight or if there's a legitimate reason for making the request to the get_items  🤔

Ultimately, I don't think it's a big deal since this is just a stop gap until WordPress 5.9 is slated for release, but I wonder if you have any thoughts on this. Emphasis on the non-blocking aspect of this comment.

creativecoder replied:
I don't have historical insight, but my guess is that it has to do with how the "id" of a template in the context of the rest API is `{stylesheet}//{template-name}`... where as "wp_id" is the ID of the template post in the database.

The `get_item` [method for the templates controller](https://github.com/WordPress/wordpress-develop/blob/ea49625d59f41e17312d4c08503cd2dd567f4bda/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L192) uses the stylesheet based template id, so calling the `get_items` route with `wp_id` is the only way to request a template by post ID right now. A workaround of sorts... it might be worth proposing that the `get_item` handler function also accept a numerical post ID.

jeyip commented on apps/editing-toolkit/editing-toolkit-plugin/dotcom-fse/templates/class-rest-templates-controller.php:
```diff
@@ -11,6 +11,28 @@
+		if ( $template_id ) {
+			$request->set_param( 'id', $template_id );
+			$template_part = self::get_item( $request )->data;
+			return \rest_ensure_response( array( $template_part ) );
```
This might be a silly question, but what does the ~forward~ back slash do in front of the `\rest_ensure_response` method call in PHP / WordPress? A quick search didn't surface any obvious answers 😅

creativecoder replied:
It's to indicate the function is in the global namespace, rather than within the namespace of the current file (`A8C\FSE`). Alternatively you can add `use function rest_ensure_response;` at the top of the file.

jeyip reviewed (commented):
Everything's looking reasonable to me. Going to test out atomic tomorrow. 👍

jeyip commented:
> I added some additional testing instructions for Atomic sites to the issue description.

I updated my original testing note comments to reflect atomic testing. I don't have much to add in the way of comments -- going to wait one more day for someone more experienced with legacy FSE to chime in, but things look good to me right now.

mattwiebe commented:
Testing notes:

1. D63129-code doesn't apply cleanly on my sandbox. I ran `install-plugin.sh etk fix/legacy-fse-with-current-gutenberg` and then manually made the same modifications to `wpcom-gutenberg-load.php` to remove the legacy version loading.
2. My navigation menu (`a8c/navigation-menu`) looks rough! Note that this is only in the template part editor, it looks ok when rendered inside the page editor.
<img width="999" alt="Screen Shot 2021-06-23 at 16 24 15" src="https://user-images.githubusercontent.com/195089/123170238-e5e5c180-d43f-11eb-8506-07280e77bb01.png"> 
3. The template part editor shows "Submit for Review" instead of "Update" (likely due to it being unlaunched). Probably not a huge deal since if you ain't launched your site by now, you're probably not gonna. 
<img width="787" alt="Screen Shot 2021-06-23 at 16 31 23" src="https://user-images.githubusercontent.com/195089/123170751-7fad6e80-d440-11eb-8a7c-23933f0ae24e.png">
4. Double inserters on the edit page view. I remember this coming up before? Is this what we fixed by pinning the legacy Gutenberg version? (Only happens on pages, posts and template parts are ok)
<img width="789" alt="Screen Shot 2021-06-23 at 16 34 35" src="https://user-images.githubusercontent.com/195089/123171095-f5193f00-d440-11eb-8359-7852132091ad.png">
5. I didn't see any block migration errors.
6. Template part editing worked 👍

creativecoder commented:
Thanks for reviewing @mattwiebe !

> D63129-code doesn't apply cleanly on my sandbox. I ran install-plugin.sh etk fix/legacy-fse-with-current-gutenberg and then manually made the same modifications to wpcom-gutenberg-load.php to remove the legacy version loading.

That's intentional: D63129-code includes the changes you'd get from the install-plugin script to simplify testing, because an additional change to wpcom Gutenberg loading is required to make this all work. You can just apply the diff (without running the script) for testing.

When this gets deployed, it'll require both running the etk script and adding the wpcom specific change.

> My navigation menu (a8c/navigation-menu) looks rough! Note that this is only in the template part editor, it looks ok when rendered inside the page editor.

It does! That's out of scope for this change, as it doesn't have to do with the Gutenberg upgrade. I think it's going to require modifying theme code to fix, but I do think we should fix it soon, as well.

> The template part editor shows "Submit for Review" instead of "Update" (likely due to it being unlaunched). Probably not a huge deal since if you ain't launched your site by now, you're probably not gonna. 

I noticed that but it seemed intermittent--I'll take a deeper look. That is confusing!

> Double inserters on the edit page view. I remember this coming up before?

I'm not sure, but I'll look into it--that should be addressed in this change if it's caused by upgrading Gutenberg.

mattwiebe commented:
> You can just apply the diff (without running the script) for testing

Sorry, I wasn't clear: I _couldn't_ apply the diff on my sandbox so I had to run the script instead. There were some ETK version mismatches in the conflict so I think it's just that the Diff needs rebasing or whatever you call that in SVN 😅 

I think we'll need to get signoff to work on the navigation issue as it's complex and we can hopefully just figure out a way to migrate those to core nav menus anyway. 

I'll dig into the double inserter today, as I found a ~month old convo about it but there were was no issue filed or action steps planned.

mattwiebe committed:
prevent double inserters while editing pages

mattwiebe commented:
I added a commit in cf2f252d2073148ad75c5272d307dba2297553a7 to prevent the double inserters while editing a page. 

The testing is simple: you should see one, and only one, inserter on any post type. Our custom inserter on pages inserts only into the `a8c/post-content` block.

Commit:
Update class for Core block inserter to hide duplicate inserter

Commit:
Modify template part controller response so it contains _links for the
template

This fixes the issue where it appears that the user doesn't have
permissions to edit the template ("Submit for Review")

creativecoder commented:
Sorry for y'all who were pinged because of the automated review requests generated by that botched rebase! Feel free to unsubscribe.

creativecoder commented:
This PR and the testing diff D63129-code are now updated.

- Fixed the double block inserter (props @mattwiebe )
- Fixed the "Submit for Review" issue

This is ready for further testing and review.

Copons commented:
I haven't tested on Atomic yet, but this looks great to me!
I've seen blocks-related console errors upon loading the editor _before applying the patch_, and those same error showed up again after the patch. But then disappeared after saving the post.
Legacy FSE works well, no more double inserter is excellent, and the code too seems spot on (beautiful idea to override `get_items` to work around the ID difference!).

jeyip commented:
The code looks reasonable to me 👍 

Walking through another round of testing to see if the new issues mentioned [here](https://github.com/Automattic/wp-calypso/pull/53612#issuecomment-867998181) were addressed.

jeyip commented:
## Testing
### Requirements
**Test block migrations when upgrading Gutenberg**
- [x] Did not see JS console errors that might have been related to block migrations for Core blocks
- [x] Did a spot check for functionality and formatting on each of the blocks

**Simple Sites: Test Compatibility of Legacy FSE Features**
- [x] Header changes can be persisted (modified site description, added text and image content)
- [x] Header changes are displayed on the front end and legacy site editor
- [x] Footer changes can be persisted (modified site description, added text and image content)
- [x] Footer changes are displayed on the front end and legacy site editor
- [x] No more "Submit for Review" text when saving template part modifications
- [x] No double block inserter on the edit page view

**Atomic Sites: Test Compatibility of Legacy FSE Features**

- [x] Header changes can be persisted (modified site description, added text and image content)
- [x] Header changes are displayed on the front end and legacy site editor
- [x] Footer changes can be persisted (modified site description, added text and image content)
- [x] Footer changes are displayed on the front end and legacy site editor
- [x] No more "Submit for Review" text when saving template part modifications
- [x] No double block inserter on the edit page view

jeyip commented:
Thanks for your stellar work on this PR @creativecoder 🌟

I think it's reasonable to handle the `a8c/navigation-menu` styling issues in a follow-up, but if you're inclined, we can continue working in this PR.

jeyip commented:
I [drafted a ticket](https://github.com/Automattic/wp-calypso/issues/54267) to track the `a8c/navigation-menu` bug.

jeyip commented:
@mattwiebe @Copons @creativecoder I smoke tested the rebased PR. Everything still appears to be working as expected. The [iframe URL bug](https://github.com/Automattic/wp-calypso/pull/54416) we were seeing should now be resolved. 

The changes here will affect live sites, so I'll wait to deploy until I sign on Monday morning.
